### DISCUSSION TITLE
CT-e 4.0

### DISCUSSION THREAD

**Discussion by hitalodiniz** [TadaSoftware/PyNFe#330] [Q&A]
Alguém já iniciou o dev para a versão do CT-e 4.0? Estou trabalhando no código para adaptar as fucionalidades da NF-e para o CT-e

**Comment by rvalyi** [comment]
Tem uma parte dos "bindings", esses objetos Python que se serializam em XML que o pyNFe faz de uma forma extramente braçal e totalmente ultrapassada que eu gero com [xsdata](https://github.com/tefra/xsdata) na lib fiscal https://github.com/akretion/nfelib . A CT-e na nfelib é a 4.0 aderente aos últimos esquemas da fazenda (para NFe, MDFe, BPe e NFSe tb alias). Vc pode usar a nfelib e depois usar a lib que vc quiser para transmitir o XML (usamos a lib erpbrasil.assinatura para assinar). Essa lib nfelib é que usamos no ERP open source Odoo que adaptamos pro Brasil na fundação OCA alias (bem diferente daquela "localização" da versão Enterprise zoada e não open source cuidado com isso).

Seria até uma sugestão pro projeto PyNfe de adotar a NFelib pros bindings Python em vez de tentar manter tudo manualmente a cada atualização sem nem saber direito até que ponto esta aderente ou não com as especificações...

**Reply by hitalodiniz** [reply]
Excelente, vou utilizar a nfelib, muito obrigado e parabéns pelo trabalho desenvolvido. E já estou avaliando o Odoo para usá-lo, que solução interessante.

**Reply by juniortada** [reply]
@hitalodiniz @rvalyi o PyNFe utilizar lxml na serialização do xml é por questão de desempenho na emissão de NFC-e.
A serialização do xml e geração do DANFE eram um gargalo no começo do projeto.
Fora isso também existe o fato de muitas UF's utilizarem customizações (principalmente para NFC-e e NFS-e) não documentadas ou mesmo bugs na implementação.
Usar lxml permite ter flexibilização para atender todas as SEFAZ e todos os tipos de notas.
Acredito que para a maioria dos usuários que utilizam somente NF-e isso nunca será uma preocupação,
mas como na época tinha clientes que precisavam principalmente de NFC-e isso era um problema.
Não sei como está o desempenho para serializar de outras formas, precisaria fazer testes, principalmente para clientes de supermercados.

**Reply by rvalyi** [reply]
Ola @juniortada o xsdata usa o lxml tb para serializar...
Não fiz benchmark porque realmente não é nenhum problema, mas pode fazer se quiser, te garanto que a serializaçao do xsdata é extremamente rapida. O próprio autor do xsdata é um cara muito ET que toma muito cuidado com isso. o xsdata já é considerada a melhor lib de databinding XML no geral (todas linguagens confundidas) e tem muita gente grande começando a usar e que faz muito benchmark e pula pro xsdata porque vê que é melhor do que tudo que usava antes... A nfelib não faz muito magia por sim so, quem faz o trabalho pesado é o xsdata mesmo.

**Comment by leogregianin** [comment]
Vou fechar essa discussão. Issue #276 aberta sobre o assunto.
